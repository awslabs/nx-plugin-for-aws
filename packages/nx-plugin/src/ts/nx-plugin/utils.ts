/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  joinPathFragments,
  readJson,
  Tree,
  updateJson,
  writeJson,
} from '@nx/devkit';
import { readProjectConfigurationUnqualified } from '../../utils/nx';
import { withVersions } from '../../utils/versions';
import PackageJson from '../../../package.json';

/**
 * Configures a TypeScript project as an Nx Plugin
 */
export const configureTsProjectAsNxPlugin = (
  tree: Tree,
  projectName: string,
) => {
  const project = readProjectConfigurationUnqualified(tree, projectName);

  const tsConfigPath = joinPathFragments(project.root, 'tsconfig.json');
  if (!tree.exists(tsConfigPath)) {
    throw new Error(
      `Selected plugin project ${projectName} is not a TypeScript project`,
    );
  }

  // Nx Plugins must use commonjs as a limitation of nx
  // https://github.com/nrwl/nx/issues/15682
  updateJson(tree, tsConfigPath, (tsConfig) => {
    tsConfig.compilerOptions ??= {};
    tsConfig.compilerOptions.module ??= 'commonjs';
    return tsConfig;
  });

  // Create an empty generators.json if one dosn't exist
  const generatorsJsonPath = joinPathFragments(project.root, 'generators.json');
  if (!tree.exists(generatorsJsonPath)) {
    writeJson(tree, generatorsJsonPath, {
      generators: {},
    });
  }

  // Create a package.json if one doesn't exist, and configure "main" and "generators"
  const pluginPackageJsonPath = joinPathFragments(project.root, 'package.json');
  if (!tree.exists(pluginPackageJsonPath)) {
    writeJson(tree, pluginPackageJsonPath, {
      name: project.name,
    });
  }
  updateJson(tree, pluginPackageJsonPath, (pkg) => {
    pkg.main ??= './src/index.js';
    pkg.generators ??= './generators.json';
    return pkg;
  });

  const rootPackageJson = tree.exists('package.json')
    ? readJson(tree, 'package.json')
    : undefined;
  const isNxPluginForAws = rootPackageJson?.name === '@aws/nx-plugin-source';

  if (!isNxPluginForAws) {
    // Add the required dependencies to the root package.json, and project's package.json
    const deps = {
      ...withVersions(['@nx/devkit']),
      // Include dependency on @aws/nx-plugin
      [PackageJson.name]: `^${PackageJson.version}`,
    };
    addDependenciesToPackageJson(tree, {}, deps);
    addDependenciesToPackageJson(tree, deps, {}, pluginPackageJsonPath);
  }
};
