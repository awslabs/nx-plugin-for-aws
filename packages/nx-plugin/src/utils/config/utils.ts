/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, joinPathFragments, Tree } from '@nx/devkit';
import { AwsNxPluginConfig } from '.';
import * as ts from 'typescript';
import { jsonToAst, replace } from '../ast';
import { factory } from 'typescript';
import { formatFilesInSubtree } from '../format';
import { importTypeScriptModule } from '../js';

export const AWS_NX_PLUGIN_CONFIG_FILE_NAME = 'aws-nx-plugin.config.mts';

/**
 * Ensure that the config file exists
 */
export const ensureAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    // Create an empty config file if it doesn't already exist
    generateFiles(tree, joinPathFragments(__dirname, 'files'), '.', {});
  }
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return await readAwsNxPluginConfig(tree)!;
};

/**
 * Read config from the aws nx plugin configuration file
 */
export const readAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig | undefined> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    return undefined;
  }
  const configTs = tree.read(AWS_NX_PLUGIN_CONFIG_FILE_NAME, 'utf-8');
  return await importTypeScriptModule(configTs);
};

/**
 * Update the aws nx plugin config file.
 * Undefined top level keys in the config update will be untouched, otherwise config is replaced
 */
export const updateAwsNxPluginConfig = async (
  tree: Tree,
  configUpdate: Partial<AwsNxPluginConfig>,
): Promise<void> => {
  // Replace the default export
  replace(
    tree,
    AWS_NX_PLUGIN_CONFIG_FILE_NAME,
    'ExportAssignment ObjectLiteralExpression',
    (node) => {
      const existingObj = node as ts.ObjectLiteralExpression;

      const existingProps = new Map<string, ts.PropertyAssignment>();
      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          existingProps.set(prop.name.getText(), prop);
        }
      });

      const properties: ts.PropertyAssignment[] = [];

      for (const [key, value] of Object.entries(configUpdate)) {
        properties.push(
          factory.createPropertyAssignment(
            key,
            jsonToAst(value) as ts.Expression,
          ),
        );
      }

      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          const name = prop.name.getText();
          if (configUpdate[name] === undefined) {
            properties.push(prop);
          }
        }
      });

      return factory.createObjectLiteralExpression(properties, true);
    },
  );

  // Format the config nicely after an update
  await formatFilesInSubtree(tree, AWS_NX_PLUGIN_CONFIG_FILE_NAME);
};
