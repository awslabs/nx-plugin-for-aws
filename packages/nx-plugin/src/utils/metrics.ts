/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree, joinPathFragments } from '@nx/devkit';
import { ArrayLiteralExpression, StringLiteral, factory } from 'typescript';
import { replaceIfExists } from './ast';
import { formatFilesInSubtree } from './format';
import { NxGeneratorInfo, getPackageVersion } from './nx';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from './shared-constructs-constants';

// Used to identify @aws/nx-plugin in AWS metrics
export const METRIC_ID = 'uksb-4wk0bqpg5s';

// File in which the MetricsAspect may exist
export const METRICS_ASPECT_FILE_PATH = joinPathFragments(
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  'src',
  'core',
  'app.ts',
);

// Query to find a particular metrics aspect variable
export const metricsAspectVariableQuery = (
  variableName: string,
  valueQuery: string,
) =>
  `ClassDeclaration[name.name="MetricsAspect"] MethodDeclaration[name.name="visit"] VariableDeclaration:has(Identifier[name="${variableName}"]) ${valueQuery}`;

/**
 * Instruments metrics by updating the MetricsAspect in common/constructs/src/core/app.ts if the file exists
 */
export const addGeneratorMetricsIfApplicable = async (
  tree: Tree,
  generatorInfo: NxGeneratorInfo[],
) => {
  if (tree.exists(METRICS_ASPECT_FILE_PATH)) {
    // Update the id
    replaceIfExists(
      tree,
      METRICS_ASPECT_FILE_PATH,
      metricsAspectVariableQuery('id', 'StringLiteral'),
      (): StringLiteral => {
        return factory.createStringLiteral(METRIC_ID, true);
      },
    );
    // Update the version
    replaceIfExists(
      tree,
      METRICS_ASPECT_FILE_PATH,
      metricsAspectVariableQuery('version', 'StringLiteral'),
      (): StringLiteral => {
        return factory.createStringLiteral(getPackageVersion(), true);
      },
    );
    // Add each generator as a tag
    replaceIfExists(
      tree,
      METRICS_ASPECT_FILE_PATH,
      metricsAspectVariableQuery('tags', 'ArrayLiteralExpression'),
      (node: ArrayLiteralExpression): ArrayLiteralExpression => {
        const existingMetrics = new Set(
          node.elements.map((element) => (element as StringLiteral)?.text),
        );

        return factory.createArrayLiteralExpression([
          ...node.elements,
          ...generatorInfo
            .filter((info) => !existingMetrics.has(info.metric))
            .map((info) => factory.createStringLiteral(info.metric, true)),
        ]);
      },
    );
    await formatFilesInSubtree(tree, METRICS_ASPECT_FILE_PATH);
  }
};
