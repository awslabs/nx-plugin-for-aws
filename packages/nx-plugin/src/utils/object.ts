/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

export const sortObjectKeys = <T>(object: {
  [key: string]: T;
}): { [key: string]: T } =>
  Object.keys(object)
    .sort()
    .reduce((obj, key) => {
      obj[key] = object[key];
      return obj;
    }, {});
