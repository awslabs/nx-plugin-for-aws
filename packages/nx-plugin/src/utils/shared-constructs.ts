/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  getPackageManagerCommand,
  joinPathFragments,
  OverwriteStrategy,
  Tree,
} from '@nx/devkit';
import { getNpmScopePrefix, toScopeAlias } from './npm-scope';
import tsProjectGenerator from '../ts/lib/generator';
import { withVersions } from './versions';
import { formatFilesInSubtree } from './format';
import {
  PACKAGES_DIR,
  TYPE_DEFINITIONS_DIR,
  TYPE_DEFINITIONS_NAME,
  SHARED_CONSTRUCTS_DIR,
  SHARED_CONSTRUCTS_NAME,
} from './shared-constructs-constants';

export async function sharedConstructsGenerator(tree: Tree) {
  const npmScopePrefix = getNpmScopePrefix(tree);
  updateGitignore(tree);

  if (
    !tree.exists(
      joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR, 'project.json'),
    )
  ) {
    await tsProjectGenerator(tree, {
      name: TYPE_DEFINITIONS_NAME,
      directory: PACKAGES_DIR,
      subDirectory: TYPE_DEFINITIONS_DIR,
    });
    tree.delete(joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR, 'src'));
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', TYPE_DEFINITIONS_DIR, 'src'),
      joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR, 'src'),
      {
        npmScopePrefix,
      },
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', 'common', 'readme'),
      joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR),
      {
        fullyQualifiedName: `${npmScopePrefix}${TYPE_DEFINITIONS_NAME}`,
        name: TYPE_DEFINITIONS_NAME,
        pkgMgrCmd: getPackageManagerCommand().exec,
      },
      {
        overwriteStrategy: OverwriteStrategy.Overwrite,
      },
    );
    await formatFilesInSubtree(tree);
  }
  if (
    !tree.exists(
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
    )
  ) {
    await tsProjectGenerator(tree, {
      name: SHARED_CONSTRUCTS_NAME,
      directory: PACKAGES_DIR,
      subDirectory: SHARED_CONSTRUCTS_DIR,
    });
    tree.delete(joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src'));
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', SHARED_CONSTRUCTS_DIR, 'src'),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src'),
      {
        npmScopePrefix,
        scopeAlias: toScopeAlias(npmScopePrefix),
      },
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', 'common', 'readme'),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR),
      {
        fullyQualifiedName: `${npmScopePrefix}${SHARED_CONSTRUCTS_NAME}`,
        name: SHARED_CONSTRUCTS_NAME,
        pkgMgrCmd: getPackageManagerCommand().exec,
      },
      {
        overwriteStrategy: OverwriteStrategy.Overwrite,
      },
    );
    addDependenciesToPackageJson(
      tree,
      withVersions(['constructs', 'aws-cdk-lib']),
      withVersions(['@types/node']),
    );
    await formatFilesInSubtree(tree);
  }
}

const updateGitignore = (tree: Tree) => {
  const gitignore = tree.exists('.gitignore')
    ? tree.read('.gitignore', 'utf-8')
    : '';
  const regex = /runtime-config.json/gm;
  const hasRuntimeConfig = regex.test(gitignore ?? '');
  if (hasRuntimeConfig) {
    return;
  }
  tree.write('.gitignore', `${gitignore}\n\nruntime-config.json`);
};
