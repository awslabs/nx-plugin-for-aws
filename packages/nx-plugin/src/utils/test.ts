/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { vi, expect } from 'vitest';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { joinPathFragments, Tree } from '@nx/devkit';
import solutionSetup from '@nx/js/src/utils/typescript/ts-solution-setup';

export const createTreeUsingTsSolutionSetup = (): Tree => {
  vi.spyOn(solutionSetup, 'isUsingTsSolutionSetup').mockImplementation(
    () => true,
  );

  const tree = createTreeWithEmptyWorkspace();

  tree.write('pnpm-workspace.yaml', `packages:\n  - packages/*`);

  tree.write('tsconfig.json', '{}');
  return tree;
};

/**
 * Snapshot all files within a directory in the given tree
 */
export const snapshotTreeDir = (tree: Tree, dir: string) => {
  if (tree.isFile(dir)) {
    expect(tree.read(dir, 'utf-8')).toMatchSnapshot(dir);
  } else {
    tree
      .children(dir)
      .forEach((subDir) =>
        snapshotTreeDir(tree, joinPathFragments(dir, subDir)),
      );
  }
};
