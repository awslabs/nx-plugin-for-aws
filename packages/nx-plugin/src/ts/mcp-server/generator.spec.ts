/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree, readJson, readProjectConfiguration } from '@nx/devkit';
import {
  tsMcpServerGenerator,
  TS_MCP_SERVER_GENERATOR_INFO,
} from './generator';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { expectHasMetricTags } from '../../utils/metrics.spec';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import { sortObjectKeys } from '../../utils/object';

describe('ts#mcp-server generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  it('should generate a TypeScript MCP server with correct structure', async () => {
    await tsMcpServerGenerator(tree, { name: 'test-server' });

    // Check that the project was created
    const config = readProjectConfiguration(tree, '@proj/test-server');
    expect(config).toBeDefined();
    expect(config.root).toBe('test-server');
    expect(config.sourceRoot).toBe('test-server/src');

    // Check that the source files were generated
    expect(tree.exists('test-server/src/index.ts')).toBeTruthy();
    expect(tree.exists('test-server/src/server.ts')).toBeTruthy();
    expect(tree.exists('test-server/src/global.d.ts')).toBeTruthy();
    expect(
      tree.exists('test-server/src/resources/example-context.md'),
    ).toBeTruthy();
    expect(tree.exists('test-server/README.md')).toBeTruthy();
  });

  it('should add required dependencies to package.json', async () => {
    await tsMcpServerGenerator(tree, { name: 'test-server' });

    // Read package.json to verify dependencies
    const packageJson = JSON.parse(tree.read('package.json', 'utf-8'));

    // Check for required dependencies
    expect(packageJson.dependencies['@modelcontextprotocol/sdk']).toBeDefined();
    expect(packageJson.dependencies['zod']).toBeDefined();

    // Check for required dev dependencies
    expect(packageJson.devDependencies['tsx']).toBeDefined();
    expect(packageJson.devDependencies['esbuild']).toBeDefined();
  });

  it('should add correct build targets to project configuration', async () => {
    await tsMcpServerGenerator(tree, { name: 'test-server' });

    const config = readProjectConfiguration(tree, '@proj/test-server');

    // Check for bundle target
    expect(config.targets.bundle).toBeDefined();
    expect(config.targets.bundle.executor).toBe('nx:run-commands');
    expect(config.targets.bundle.options.commands[0]).toContain('esbuild');
    expect(config.targets.bundle.options.commands[0]).toContain('--bundle');
    expect(config.targets.bundle.options.commands[0]).toContain(
      '--platform=node',
    );
    expect(config.targets.bundle.options.commands[0]).toContain('--format=esm');
    expect(config.targets.bundle.options.commands[0]).toContain(
      '--loader:.md=text',
    );

    // Check for dev target
    expect(config.targets.dev).toBeDefined();
    expect(config.targets.dev.executor).toBe('nx:run-commands');
    expect(config.targets.dev.options.commands[0]).toContain('nx watch');
    expect(config.targets.dev.options.commands[0]).toContain(
      'nx run @proj/test-server:bundle',
    );

    // Check that build depends on bundle
    expect(config.targets.build).toBeDefined();
    expect(config.targets.build.dependsOn).toContain('bundle');
  });

  it('should handle custom directory path', async () => {
    await tsMcpServerGenerator(tree, {
      name: 'test-server',
      directory: 'custom-dir',
    });

    // Check that the project was created in the custom directory
    const config = readProjectConfiguration(tree, '@proj/test-server');
    expect(config).toBeDefined();
    expect(config.root).toBe('custom-dir/test-server');
    expect(config.sourceRoot).toBe('custom-dir/test-server/src');

    // Check that the source files were generated in the custom directory
    expect(tree.exists('custom-dir/test-server/src/index.ts')).toBeTruthy();
    expect(tree.exists('custom-dir/test-server/src/server.ts')).toBeTruthy();
  });

  it('should add generator to project metadata', async () => {
    // Call the generator function
    await tsMcpServerGenerator(tree, { name: 'test-server' });

    expect(readJson(tree, 'test-server/project.json').metadata).toHaveProperty(
      'generator',
      TS_MCP_SERVER_GENERATOR_INFO.id,
    );
  });

  it('should match snapshot', async () => {
    await tsMcpServerGenerator(tree, { name: 'test-server' });

    const changes = sortObjectKeys(
      tree
        .listChanges()
        .filter((f) => f.path.startsWith('test-server/'))
        .reduce((acc, curr) => {
          acc[curr.path] = tree.read(curr.path, 'utf-8');
          return acc;
        }, {}),
    );

    // Verify project files
    expect(changes).toMatchSnapshot('mcp-server-files');
  });

  it('should add generator metric to app.ts', async () => {
    await sharedConstructsGenerator(tree);

    await tsMcpServerGenerator(tree, { name: 'test-server' });

    expectHasMetricTags(tree, TS_MCP_SERVER_GENERATOR_INFO.metric);
  });
});
