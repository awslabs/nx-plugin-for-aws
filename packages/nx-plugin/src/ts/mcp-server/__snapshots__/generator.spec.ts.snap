// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ts#mcp-server generator > should match snapshot > mcp-server-files 1`] = `
{
  "test-server/README.md": "# test-server

This package defines a Model Context Protocol (MCP) Server in TypeScript.

## What is MCP?

The Model Context Protocol (MCP) is a standardized protocol for providing context to Large Language Models (LLMs). It allows AI assistants to access external tools and resources through a consistent interface.

## Documentation

- [Model Context Protocol Documentation](https://modelcontextprotocol.io/)
- [MCP TypeScript SDK Documentation](https://github.com/modelcontextprotocol/sdk-typescript)

## Developer Guide

### Getting Started

This MCP server is built using the MCP TypeScript SDK. Here's how to work with it:

1. **Server Structure**: The main server code is in \`src/server.ts\`. This file defines the server, its tools, and resources.

2. **Adding Tools**: Tools are functions that the LLM can call to perform actions. Add new tools using the \`server.tool()\` method:

\`\`\`typescript
server.tool(
  'toolName',
  { param1: z.string(), param2: z.number() }, // Input schema using Zod
  async ({ param1, param2 }) => {
    // Tool implementation
    return {
      content: [{ type: 'text', text: 'Result' }],
    };
  },
);
\`\`\`

3. **Adding Resources**: Resources provide context to the LLM. You can add static resources from files or dynamic resources:

\`\`\`typescript
// Static resource from a file
import resourceContent from './resources/my-resource.md';

server.resource('resource-name', 'example://resource', async (uri) => ({
  contents: [{ uri: uri.href, text: resourceContent }],
}));

// Dynamic resource
server.resource('dynamic-resource', 'dynamic://resource', async (uri) => {
  const data = await fetchSomeData();
  return {
    contents: [{ uri: uri.href, text: data }],
  };
});
\`\`\`

### Building the Server

To build the server for use with AI assistants:

\`\`\`bash
nx run <project-name>:bundle
\`\`\`

This creates a bundled version of your server in \`dist/<project-path>/bundle/index.js\`.

## Using with AI Assistants

To use your MCP server with AI assistants, you need to build it first to create the bundle:

\`\`\`bash
nx run <project-name>:bundle
\`\`\`

### Configuration

Most AI assistants (Amazon Q Developer CLI, Cline, Cursor, Claude Code, etc.) use a similar configuration approach. You'll need to create or update a configuration file with your MCP server details:

\`\`\`json
{
  "mcpServers": {
    "test-server": {
      "command": "node",
      "args": ["/path/to/workspace/dist/test-server/bundle/index.js"],
      "transportType": "stdio"
    }
  }
}
\`\`\`

Replace \`/path/to/workspace/dist/test-server/bundle/index.js\` with the actual path to your bundled MCP server.

Note that if you receive an error due to \`node\` missing (eg \`ENOENT node\`), you might need to specify the full path to \`node\` which you can obtain by running \`which node\`.

### Development Mode

During development, you can use the \`dev\` target to continuously rebuild the bundle whenever you make changes:

\`\`\`bash
nx run <project-name>:dev
\`\`\`

This will watch for changes in your project files and automatically rebuild the bundle.

Whenever you've made changes, you'll need to restart the MCP server in your AI assistant to test it out. The exact process depends on the assistant, but generally:

1. Find your MCP server in the assistant's settings or configuration
2. Look for a "Restart" or "Reload" option
",
  "test-server/eslint.config.mjs": "import baseConfig from '../eslint.config.mjs';

export default [
  ...baseConfig,
  {
    files: ['**/*.json'],
    rules: {
      '@nx/dependency-checks': [
        'warn',
        {
          ignoredFiles: [
            '{projectRoot}/eslint.config.{js,cjs,mjs}',
            '{projectRoot}/vite.config.{js,ts,mjs,mts}',
          ],
        },
      ],
    },
    languageOptions: {
      parser: await import('jsonc-eslint-parser'),
    },
  },
];
",
  "test-server/project.json": "{
  "name": "@proj/test-server",
  "$schema": "../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "test-server/src",
  "projectType": "library",
  "tags": [],
  "metadata": {
    "generator": "ts#mcp-server"
  },
  "targets": {
    "build": {
      "dependsOn": ["lint", "compile", "test", "bundle"]
    },
    "compile": {
      "executor": "nx:run-commands",
      "outputs": ["{workspaceRoot}/dist/test-server/tsc"],
      "options": {
        "command": "tsc --build tsconfig.lib.json",
        "cwd": "{projectRoot}"
      }
    },
    "test": {
      "executor": "@nx/vite:test",
      "outputs": ["{options.reportsDirectory}"],
      "options": {
        "reportsDirectory": "../coverage/test-server"
      }
    },
    "bundle": {
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          "esbuild \\"test-server/src/index.ts\\" --bundle --platform=node --format=esm --loader:.md=text --outfile=dist/test-server/bundle/index.js"
        ]
      }
    },
    "dev": {
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          "nx watch --projects=@proj/test-server --includeDependentProjects -- nx run @proj/test-server:bundle"
        ]
      }
    }
  }
}
",
  "test-server/src/global.d.ts": "declare module '*.md' {
  const content: string;
  export default content;
}
",
  "test-server/src/index.ts": "import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { createServer } from './server.js';

// Start receiving messages on stdin and sending messages on stdout
const transport = new StdioServerTransport();
await createServer().connect(transport);
",
  "test-server/src/resources/example-context.md": "## Example Context

This is some example context in a markdown file!
",
  "test-server/src/server.ts": "import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import exampleContext from './resources/example-context.md';

/**
 * Create the MCP Server
 */
export const createServer = () => {
  const server = new McpServer({
    name: 'test-server',
    version: '1.0.0',
  });

  // Add an addition tool
  server.tool(
    'add',
    'adds two numbers',
    { a: z.number(), b: z.number() },
    async ({ a, b }) => ({
      content: [{ type: 'text', text: String(a + b) }],
    }),
  );

  // Add a resource which provides context from a markdown file
  server.resource('example-context', 'example://context', async (uri) => ({
    contents: [{ uri: uri.href, text: exampleContext }],
  }));

  return server;
};
",
  "test-server/tsconfig.json": "{
  "extends": "../tsconfig.base.json",
  "files": [],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ],
  "compilerOptions": {}
}
",
  "test-server/tsconfig.lib.json": "{
  "extends": "../tsconfig.base.json",
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "../dist/test-server/tsc",
    "tsBuildInfoFile": "../dist/test-server/tsc/tsconfig.lib.tsbuildinfo",
    "emitDeclarationOnly": false,
    "module": "nodenext",
    "moduleResolution": "nodenext",
    "types": ["node"]
  },
  "include": ["src/**/*.ts"],
  "references": [],
  "exclude": [
    "vite.config.ts",
    "vite.config.mts",
    "vitest.config.ts",
    "vitest.config.mts",
    "src/**/*.test.ts",
    "src/**/*.spec.ts",
    "src/**/*.test.tsx",
    "src/**/*.spec.tsx",
    "src/**/*.test.js",
    "src/**/*.spec.js",
    "src/**/*.test.jsx",
    "src/**/*.spec.jsx"
  ]
}
",
  "test-server/tsconfig.spec.json": "{
  "extends": "../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./out-tsc/vitest",
    "types": [
      "vitest/globals",
      "vitest/importMeta",
      "vite/client",
      "node",
      "vitest"
    ],
    "module": "nodenext",
    "moduleResolution": "nodenext"
  },
  "include": [
    "vite.config.ts",
    "vite.config.mts",
    "vitest.config.ts",
    "vitest.config.mts",
    "src/**/*.test.ts",
    "src/**/*.spec.ts",
    "src/**/*.test.tsx",
    "src/**/*.spec.tsx",
    "src/**/*.test.js",
    "src/**/*.spec.js",
    "src/**/*.test.jsx",
    "src/**/*.spec.jsx",
    "src/**/*.d.ts"
  ],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    }
  ]
}
",
  "test-server/vite.config.ts": "import { defineConfig } from 'vite';

export default defineConfig(() => ({
  root: __dirname,
  cacheDir: '../node_modules/.vite/test-server',
  plugins: [],
  // Uncomment this if you are using workers.
  // worker: {
  //  plugins: [ nxViteTsPaths() ],
  // },
  test: {
    watch: false,
    globals: true,
    environment: 'jsdom',
    include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
    reporters: ['default'],
    coverage: {
      reportsDirectory: './test-output/vitest/coverage',
      provider: 'v8' as const,
    },
    passWithNoTests: true,
  },
}));
",
}
`;
